mmeval/metrics/average_precision.py

```diff
@@ -0,0 +1,384 @@
+    precision = tps / pred_pos_nums.unsqueeze(-1).float()  # divide along rows
+    ap = flow.sum(precision, 0) / flow.clamp(total_pos, min=1)
+
+    if average == 'macro':
+        return ap.mean() * 100.0
+    else:
+        return ap * 100
+
+
+def _average_precision(preds: np.ndarray, labels: np.ndarray,
+                       average) -> np.ndarray:
+    r"""Calculate the average precision for numpy.
+
+    AP summarizes a precision-recall curve as the weighted mean of maximum
+    precisions obtained for any r'>r, where r is the recall:
+
+    .. math::
+        \text{AP} = \sum_n (R_n - R_{n-1}) P_n
+
+    Note that no approximation is involved since the curve is piecewise
+    constant.
+
+    Args:
+        preds (np.ndarray): The model prediction with shape
+            ``(N, num_classes)``.
+        labels (np.ndarray): The target of predictions with shape
+            ``(N, num_classes)``.
+
+    Returns:
+        np.ndarray: average precision result.
+    """
+    # sort examples along classes
+    sorted_pred_inds = np.argsort(-preds, axis=0)
+    sorted_target = np.take_along_axis(labels, sorted_pred_inds, axis=0)
+
+    # get indexes when gt_true is positive
+    pos_inds = sorted_target == 1
+
+    # Calculate cumulative tp case numbers
+    tps = np.cumsum(pos_inds, 0)
+    total_pos = tps[-1].copy()  # the last of tensor may change later
+
+    # Calculate cumulative tp&fp(pred_poss) case numbers
+    pred_pos_nums = np.arange(1, len(sorted_target) + 1)
+
+    tps[np.logical_not(pos_inds)] = 0
+    precision = np.divide(
+        tps, np.expand_dims(pred_pos_nums, -1), dtype=np.float32)
+    ap = np.divide(
+        np.sum(precision, 0), np.clip(total_pos, 1, np.inf), dtype=np.float32)
+
+    if average == 'macro':
+        return ap.mean() * 100.0
+    else:
+        return ap * 100
+
+
+class AveragePrecision(MultiLabelMixin, BaseMetric):
+    """Calculate the average precision with respect of classes.
+
+    Args:
+        average (str, optional): The average method. It supports two modes:
+
+            - `"macro"`: Calculate metrics for each category, and calculate
+                the mean value over all categories.
+            - `None`: Return scores of all categories.
+
+        Defaults to "macro".
+
+    References
+    ----------
+    .. [1] `Wikipedia entry for the Average precision
+           <https://en.wikipedia.org/w/index.php?title=Information_retrieval&
+           oldid=793358396#Average_precision>`_
+
+    Examples:
+
+        >>> from mmeval import AveragePrecision
+        >>> average_precision = AveragePrecision()
+
+    Use Builtin implementation with label-format labels:
+
+        >>> preds = [[0.9, 0.8, 0.3, 0.2],
+                     [0.1, 0.2, 0.2, 0.1],
+                     [0.7, 0.5, 0.9, 0.3],
+                     [0.8, 0.1, 0.1, 0.2]]
+        >>> labels = [[0, 1], [1], [2], [0]]
+        >>> average_precision(preds, labels)
+        {'mAP': 70.833..}
```

Comment:
Just to clarify: is it a common case that `mAP` is represented as a percentage number? I've seen cases like `0.70833` before and not sure which is the convention.

Reply:
Both representations should be fine, some papers use percentage as well, refers to [here](https://www.researchgate.net/publication/311668581_The_More_You_Know_Using_Knowledge_Graphs_for_Image_Classification).
And for consistency with other metrics, such as acc, precision, recall, which all use the percentage format.